packages/store/src/database/backends/indexedDB.ts

```diff
@@ -1,48 +1,304 @@
+          ).primaryKeys();
+
+          for (const [derivedPrimaryKey] of derivedRowsPrimaryKeys) {
+            includedTaskIDs.add(derivedPrimaryKey as TaskID);
+          }
+      }
+
+      // was the afterID specified at the same time?
+      if (query.afterID) {
+        // simulate a join across the two tables
+        const afterRowID = await this._fetchPrimaryKeyFromUniqueKey(
+          this.db.entities,
+          DexieEventKeys.ID,
+          query.afterID,
+        );
+
+        // Note that our approach here involves fetching _all_ matching entity IDs, then skipping the
+        // ones before the requested row. This means pagination over N entities with a predicate has
+        // quadratic time cost.
+        baseQuery = this.db.entities
+          .where(DexieEntityKeys.RowID)
+          .above(afterRowID)
+          .filter((row) => includedTaskIDs.has(row.id));
+      } else {
+        baseQuery = this.db.entities
+          .where(DexieEntityKeys.ID)
+          .anyOf([...includedTaskIDs]);
+      }
+    } else if (query.afterID) {
+      const afterRowID = await this._fetchPrimaryKeyFromUniqueKey(
+        this.db.entities,
+        DexieEventKeys.ID,
+        query.afterID,
+      );
+      baseQuery = this.db.entities
+        .where(DexieEntityKeys.RowID)
+        .above(afterRowID);
+    } else {
+      baseQuery = this.db.entities.toCollection();
+    }
+
+    if (query.limit) {
+      baseQuery = baseQuery.limit(query.limit);
+    }
+
+    const queriedEntities = await baseQuery.toArray();
+    return queriedEntities.map((entity) => ({
+      lastEventID: entity.lastEventID,
+      entity: JSON.parse(entity.data),
+    }));
   }
 
-  listEvents(query: DatabaseEventQuery): Promise<Event[]> {
-    return Promise.resolve([]);
+  async listEvents(query: DatabaseEventQuery): Promise<Event[]> {
+    let baseQuery: Dexie.Collection<DexieEventRow, number>;
+    if (query.predicate && query.afterID) {
+      const afterSequenceNumber = await this._fetchPrimaryKeyFromUniqueKey(
+        this.db.events,
+        DexieEventKeys.ID,
+        query.afterID,
+      );
+
+      // both the predicate and afterID were specified need to do more complex querying
+      if (query.predicate[1] == "=") {
+        // can create a compound query in this case
+        baseQuery = this.db.events
+          .where(`[${query.predicate[0]}+${DexieEventKeys.SequenceNumber}]`)
+          .between(
+            [query.predicate[2], afterSequenceNumber + 1],
+            [query.predicate[2], Dexie.maxKey],
+          );
+      } else {
+        // arbitrary comparison query
+        const predicatedQuery: WhereClause<
+          DexieEventRowWithPrimaryKey,
+          number
+        > = this.db.events.where(query.predicate[0]);
+        baseQuery = compareUsingPredicate(
+          predicatedQuery,
+          query.predicate,
+        ).filter(
+          (item) => item[DexieEventKeys.SequenceNumber] > afterSequenceNumber,
+        );
```

Comment:
We do indeed still need the `sortBy`, since we're querying an index, which is naturally ordered by the index's key (e.g. the event type string or whatever). And for that matter, we do still need it at line 182 also (sorry if I told you otherwise!). It's a shame our unit tests didn't catch that.

We _don't_ need it on line 166 because that index's final column is the sequence number. And we don't need it on 192 because we're using the primary index, which is already sorted by sequence number. I suppose these two cases _are_ a bit "magic"—perhaps comments in those two spots about why our approach returns the correct order?

Reply:
Gotcha! Done. I needed to refactor the control flow quite a bit since `sortedBy` returns a promise rather than a collection like the other modifiers. This involved constructing a promise in each of the branches with `toArray`. The only downside with this approach is that each of the branches now needed to the  `.limit` modifier. This got tricky since limit does not accept an optional. Instead, I created a new wrapper function to conditionally apply the limit. Let me know if you have a more clear control flow in mind!

Note: doing something like `.limit(query.limt ?? Number.MAX_VALUE)` results in a [thrown error from fake-indexeddb](https://github.com/dumbmatter/fakeIndexedDB/blob/master/src/lib/enforceRange.ts#L11).